whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,213 @@
+import net.ripe.db.whois.common.rpsl.RpslObject;
+import net.ripe.db.whois.common.rpsl.attrs.Domain;
+import net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus;
+import net.ripe.db.whois.common.rpsl.attrs.InetnumStatus;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus.ALLOCATED_BY_RIR;
+import static net.ripe.db.whois.common.rpsl.attrs.InetnumStatus.ALLOCATED_UNSPECIFIED;
+
+
+@Service
+public class RdapRelationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RdapRelationService.class);
+
+    private final Ipv4Tree ip4Tree;
+    private final Ipv6Tree ip6Tree;
+    private final Ipv4DomainTree ipv4DomainTree;
+    private final Ipv6DomainTree ipv6DomainTree;
+    private final RpslObjectDao rpslObjectDao;
+
+    @Autowired
+    public RdapRelationService(final Ipv4Tree ip4Tree, final Ipv6Tree ip6Tree,
+                               final Ipv4DomainTree ipv4DomainTree, final Ipv6DomainTree ipv6DomainTree,
+                               final RpslObjectDao rpslObjectDao) {
+        this.ip4Tree = ip4Tree;
+        this.ip6Tree = ip6Tree;
+        this.ipv4DomainTree = ipv4DomainTree;
+        this.ipv6DomainTree = ipv6DomainTree;
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    public List<String> getDomainRelationPkeys(final String pkey, final RelationType relationType){
+        final Domain domain = Domain.parse(pkey);
+        final IpInterval reverseIp = domain.getReverseIp();
+        final List<IpEntry> ipEntries = getIpEntries(getIpDomainTree(reverseIp), relationType, reverseIp);
+
+        return ipEntries
+                .stream()
+                .map(ipEntry -> rpslObjectDao.getById(ipEntry.getObjectId()).getKey().toString())
+                .toList();
+    }
+
+    public List<String> getInetnumRelationPkeys(final String pkey, final RelationType relationType){
+        final IpInterval ip = IpInterval.parse(pkey);
+        final List<IpEntry> ipEntries = getIpEntries(getIpTree(ip), relationType, ip);
+        return ipEntries
+                .stream()
+                .map(ipEntry -> ipEntry.getKey().toString())
+                .toList();
+    }
+
+    private List<IpEntry> getIpEntries(final IpTree ipTree, final RelationType relationType,
+                                       final IpInterval reverseIp) {
+        return switch (relationType) {
+            case UP -> List.of(searchFirstLessSpecific(ipTree, reverseIp));
+            case TOP -> searchTopLevelResource(ipTree, reverseIp);
+            case DOWN -> ipTree.findFirstMoreSpecific(reverseIp);
+            case BOTTOM -> searchMostSpecificFillingOverlaps(ipTree, reverseIp);
```

Comment:
Correct me if I am wrong. For BOTTOM objects : FindAllMoreSpecific and then for each entry in it findFirstMoreSpecific. If the size of findFirstMoreSpecific is less than 2 you add it in the list else ignore

Reply:
talked over a call.
The amount of the children doesn't matter. We need to check if childrens covers the father.
If we go from up to bottom...we need to iterate all of them until we get the last leaf. We don't know if we can add the node or not until we process all the leafs.
If we go from bottom to up...we are already iterating the leaf, so we just need to add it